File: solutions/93/main.py
class Solution:
    def __init__(self): 
        self.path = []
        self.result = []
    @staticmethod 
    def isValid(s, start, end): 
        if start > end: 
            return False 
        if s[start] == '0' and start != end: 
            return False 
        num = 0
        for i in range(start, end+1):
            if s[i] > '9' or s[i] < '0':
                return False 
            n = int(s[i])
            num = num*10 + n 
            if num > 255: 
                return False 
        return True 
    def backtracking(self, s, startId): 
        # base case 
        if len(self.path) == 4 and startId == len(s):
            self.result.append(".".join(self.path[:]))
            return 
        for i in range(startId, len(s)):
            if self.isValid(s, startId, i): 
                self.path.append(s[startId:i+1])
                self.backtracking(s, i+1)
                self.path.pop()
            else:
                break 
    def restoreIpAddresses(self, s):
        if len(s) > 12:
            return self.result 
        self.backtracking(s, 0)
        return self.result 

if __name__ == "__main__":
    s = Solution()
    myStr = "25525511135"
    print(s.restoreIpAddresses(myStr))

File: solutions/93/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Wed Nov  3 16:40:47 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <iostream> 
#include <string> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    vector<string> result;
    string path;
public:
    bool isValid(string& s, int start, int end) {
        if (start > end) return false;
        // string begins with 0 is invalid
        if (s[start] == '0' && start != end) return false;
        int num = 0;
        for (int i = start; i <= end; ++i) {
            // has to be digits
            if (s[i] < '0' || s[i] > '9') return false;
            int n = s[i] - '0';
            num = num * 10 + n;
            // has to be smaller than 255
            if (num > 255) return false;
        }
        return true;
    }
    void backtracking(string& s, int startId, int pointNum) {
        // base case
        if (pointNum == 3) {
            if (isValid(s, startId, s.size()-1)) {
                const int preSize = path.size();
                path += s.substr(startId, s.size()-startId+1);
                result.push_back(path);
                path.erase(preSize, s.size()-startId+1);
            }
            return;
        }
        for (int i = startId; i < s.size(); ++i) {
            if (isValid(s, startId, i)) {
                const int preSize = path.size();
                path += s.substr(startId, i-startId+1) + ".";
                pointNum++;
                backtracking(s, i+1, pointNum);
                pointNum--;
                path.erase(preSize, i-startId+2);
            }
            // note, just end this level here
            else break;
        }
    }
    vector<string> restoreIpAddresses(string s) {
        if (s.size() > 12) return result;
        backtracking(s, 0, 0);
        return result;
    }
};

TEST(Test93, SimpleTest) {
    string myStr = "25525511135";
    Solution s;
    vector<string> result = s.restoreIpAddresses(myStr);
    for (auto& r : result)
        cout << r << endl; 
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
